mesonbuild/cargo/interpreter.py

```diff
@@ -0,0 +1,449 @@
+
+# tomllib is present in python 3.11, before that it is a pypi module called tomli,
+# we try to import tomllib, then tomli,
+# TODO: add a fallback to toml2json?
+tomllib: T.Optional[ModuleType] = None
+toml2json: T.Optional[str] = None
+for t in ['tomllib', 'tomli']:
+    try:
+        tomllib = importlib.import_module(t)
+        break
+    except ImportError:
+        pass
+else:
+    # TODO: it would be better to use an Executable here, which could be looked
+    # up in the cross file or provided by a wrap. However, that will have to be
+    # passed in externally, since we don't have (and I don't think we should),
+    # have access to the `Environment` for that in this module.
+    toml2json = shutil.which('toml2json')
+
+
+def load_toml(filename: str) -> T.Dict[object, object]:
+    if tomllib:
+        with open(filename, 'rb') as f:
+            raw = tomllib.load(f)
+    else:
+        if toml2json is None:
+            raise MesonException('Could not find an implementation of tomllib, nor toml2json')
+
+        p, out, err = Popen_safe([toml2json, filename])
+        if p.returncode != 0:
+            raise MesonException('toml2json failed to decode output\n', err)
+
+        raw = json.loads(out)
+
+    if not isinstance(raw, dict):
+        raise MesonException("Cargo.toml isn't a dictionary? How did that happen?")
+
+    return raw
+
+
+def fixup_meson_varname(name: str) -> str:
+    """Fixup a meson variable name
+
+    :param name: The name to fix
+    :return: the fixed name
+    """
+    return name.replace('-', '_')
+
+# Pylance can figure out that these do not, in fact, overlap, but mypy can't
+@T.overload
+def _fixup_raw_mappings(d: manifest.BuildTarget) -> manifest.FixedBuildTarget: ...  # type: ignore
+
+@T.overload
+def _fixup_raw_mappings(d: manifest.LibTarget) -> manifest.FixedLibTarget: ...  # type: ignore
+
+@T.overload
+def _fixup_raw_mappings(d: manifest.Dependency) -> manifest.FixedDependency: ...
+
+@T.overload
+def _fixup_raw_mappings(d: manifest.Package) -> manifest.FixedPackage: ...
+
+def _fixup_raw_mappings(d: T.Union[manifest.BuildTarget, manifest.LibTarget, manifest.Dependency, manifest.Package]
+                        ) -> T.Union[manifest.FixedBuildTarget, manifest.FixedLibTarget,
+                                     manifest.FixedDependency, manifest.FixedPackage]:
+    """Fixup raw cargo mappings to ones more suitable for python to consume.
+
+    This does the following:
+    * replaces any `-` with `_`, cargo likes the former, but python dicts make
+      keys with `-` in them awkward to work with
+    * Convert Dependndency versions from the cargo format to something meson
+      understands
+
+    :param d: The mapping to fix
+    :return: the fixed string
+    """
+    raw = {fixup_meson_varname(k): v for k, v in d.items()}
+    if 'version' in raw:
+        assert isinstance(raw['version'], str), 'for mypy'
+        raw['version'] = version.convert(raw['version'])
```

Comment:
I started testing that code, and found a bug here. That changes the type of version from `str` to `List[str]`. It is later passed to `Package` constructor in `_convert_manifest()`. Mypy did not detect that issue, but that causes down the road to create a `StringNode` with a list value instead of str.

Reply:
Actually, the actual problem is `project( version: )` kwarg cannot be a list. We should not call `version.convert()` in that case.

Reply:
Suggestion that fix the issue:

```diff
diff --git a/mesonbuild/cargo/interpreter.py b/mesonbuild/cargo/interpreter.py
index a79cbf882..56e465b02 100644
--- a/mesonbuild/cargo/interpreter.py
+++ b/mesonbuild/cargo/interpreter.py
@@ -310,8 +310,10 @@ def _convert_manifest(raw_manifest: manifest.Manifest, subdir: str, path: str =
     # including if other fields are set in the lib section
     lib.setdefault('name', raw_manifest['package']['name'])
 
+    pkg = {fixup_meson_varname(k): v for k, v in raw_manifest['package'].items()}
+
     return Manifest(
-        Package(**_fixup_raw_mappings(raw_manifest['package'])),
+        Package(**pkg),
         {k: Dependency.from_raw(v) for k, v in raw_manifest.get('dependencies', {}).items()},
         {k: Dependency.from_raw(v) for k, v in raw_manifest.get('dev-dependencies', {}).items()},
         {k: Dependency.from_raw(v) for k, v in raw_manifest.get('build-dependencies', {}).items()},
```

Reply:
That isn't sufficient, we still have to cast to get pkg to be `FixedManifest` for mypy

Reply:
Btw, shouldn't we have a `FixedManifest` class that have fields like `'package': FixedPackage``?

Reply:
Nevermind, I understood that's what interpreter.Manifest is, not the same thing as manifest.Manifest.

mypy is happy with `pkg = T.cast('manifest.FixedPackage', {fixup_meson_varname(k): v for k, v in raw_manifest['package'].items()})` indeed.

Reply:
@dcbaker AFAIK this is the last blocker, right?

Reply:
I guess? This is fixed in the latest version

Reply:
Oh missed that you pushed a fix already. Thanks